src/Microsoft.Health.Dicom.Core/Features/CustomTag/IndexTagService.cs

```diff
@@ -0,0 +1,73 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dicom;
+using EnsureThat;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Core.Configs;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Query;
+
+namespace Microsoft.Health.Dicom.Core.Features.CustomTag
+{
+    public class IndexTagService : IIndexTagService
+    {
+        private readonly ICustomTagStore _customTagStore;
+        private readonly bool _enableCustomQueryTags;
+        public static readonly IReadOnlyList<IndexTag> CoreIndexTags = GetCoreIndexTags();
+        private List<IndexTag> _allIndexTags;
+        private int _allIndexTagsStatus;
+        private TaskCompletionSource<bool> _allIndexTagsCompletionSource = new TaskCompletionSource<bool>();
+
+        public IndexTagService(ICustomTagStore customTagStore, IOptions<FeatureConfiguration> featureConfiguration)
+        {
+            EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
+            EnsureArg.IsNotNull(featureConfiguration?.Value, nameof(featureConfiguration));
+            _customTagStore = customTagStore;
+            _enableCustomQueryTags = featureConfiguration.Value.EnableCustomQueryTags;
+        }
+
+        public async Task<IReadOnlyCollection<IndexTag>> GetIndexTagsAsync(CancellationToken cancellationToken = default)
+        {
+            if (_enableCustomQueryTags)
+            {
+                if (Interlocked.CompareExchange(ref _allIndexTagsStatus, 1, 0) == 0)
+                {
+                    _allIndexTags = new List<IndexTag>(CoreIndexTags);
+
+                    IReadOnlyList<CustomTagStoreEntry> customTagEntries = await _customTagStore.GetCustomTagsAsync(cancellationToken: cancellationToken);
+                    foreach (CustomTagStoreEntry customTagEntry in customTagEntries)
```

Comment:
can you move this conversion to a method on IndexTag.cs
static IndexTag Convert(CustomTagStoreEntry)